FlowCal/plot.py

```diff
@@ -1061,6 +1302,1271 @@ def hist1d(data_list,
+    using vertical, normalized, symmetrical histograms ("violins") centered on
+    corresponding x-axis values. Wider regions of violins indicate regions
+    that occur with greater frequency.
+
+    Parameters
+    ----------
+    data : 1D or ND sequence or sequence of 1D or ND sequences
+        A population or collection of populations for which to plot violins.
+        If ND sequences are used (e.g. FCSData), `channel` must be specified.
+    channel : int or str, optional
+        Channel from `data` to plot. If specified, data are assumed to be ND
+        sequences. String channel specifications are only supported for data
+        types which support string-based indexing (e.g. FCSData).
+    positions : scalar or sequence of scalars, optional
+        Positions (x-axis values) at which to center violins.
+    min_data : sequence of scalars, optional
+        A population representing a minimum control. This violin is separately
+        illustrated at the left of the plot.
+    max_data : sequence of scalars, optional
+        A population representing a maximum control. This violin is separately
+        illustrated at the left of the plot.
+    logx_zero_data : sequence of scalars, optional
+        A population representing position=0 if `xscale` is 'log'. This violin
+        is separately illustrated at the left of the plot. Ignored if `xscale`
+        is not 'log'.
+    violin_width : scalar, optional
+        Width of violin. If `xscale` is 'log', the units are decades. If not
+        specified, `violin_width` is calculated from `data_xlim` and
+        `violin_width_to_span_fraction`. If only one violin is specified in
+        `data`, `violin_width` = 0.5.
+    savefig : str, optional
+        The name of the file to save the figure to. If None, do not save.
+
+    Other parameters
+    ----------------
+    xscale : {'linear','log'}, optional
+        Scale of the x-axis.
+    yscale : {'linear','log'}, optional
+        Scale of the y-axis.
+    data_xlim : 2-element sequence, optional
+        Limits of the x-axis where `data` are plotted. If min, max, and zero
+        violins are not used, `data_xlim` also specifies plt.xlim(). If not
+        specified, `data_lim` is calculated to pad leftmost and rightmost
+        violins with 0.5*`violin_width`. If `violin_width` is also not
+        specified, `violin_width` is calculated to satisfy the
+        0.5*`violin_width` padding and `violin_width_to_span_fraction`.
+    ylim : 2-element sequence, optional
+        Limits of the y-axis. If not specified, `ylim` is calculated to span
+        all violins (before aesthetic trimming).
+    num_y_bins : int, optional
+        Number of bins to bin population members into along the y-axis.
+        Ignored if `y_bin_edges` is specified.
+    y_bin_edges : sequence of scalars, sequence of sequence of scalars, or
+                  mapping to sequence of scalars, optional
```

Comment:
Do we expect any mapping other than dictionaries? Dictionaries appear to be the only standard mapping in python. I would suggest simplifying the terminology to "dictionary".

Reply:
Same with "sequence" -> "list", unless we're expecting an array, for which there is the "array_like" term. This would make it more consistent with the rest of `Flowcal.plot`.

Reply:
I don't know what you mean by "expect". If robust, the function should work with any reasonable input. My nomenclature is intended to be as general as possible and is inspired by Python's [Abstract Base Classes (ABCs)](https://docs.python.org/3/library/collections.abc.html), which include `Sequence` and `Mapping`. Moreover, the `Sequence` and `Mapping` ABCs are used pretty extensively throughout the code. So I guess you could say the code expects anything that is a `Sequence` or a `Mapping`.

More concretely, `Sequence` includes `list` _and_ `tuple`, and `Mapping` includes `dict` and `OrderedDict`.

It appears `ndarray` doesn't quite make the cut for `Sequence`: https://github.com/numpy/numpy/issues/2776, https://github.com/numpy/numpy/issues/7315. This could actually be an issue in a few places (e.g. specifying `upper_trim_fraction` or `lower_trim_fraction` as an `ndarray` of `floats`); I didn't test `ndarrays` extensively.

And `array_like` can actually be quite vague: https://stackoverflow.com/a/40380014.

Ultimately, I wanted `tuple` to work as well as `list` and `ndarray` anywhere I said "sequence". `pd.Series` and `pd.DataFrame` would also be nice, but I consider those the purview of https://github.com/taborlab/FlowCal/issues/76.

Reply:
I understand that you're trying to be general with these abstract classes, but these `Sequence` and `Mapping` names are not widespread and  make the documentation more confusing than it needs to be. I didn't even know of their existence, and only knew where to look them up after seeing that they come from `collections` after looking at the code. Furthermore, these are not used anywhere else in `FlowCal`, and it is very inconsistent to have one single function that suddenly uses this nomenclature. I'd rather use `list` and `dictionary` and drop the theoretical compatibility with other `Sequence` and `Mapping` objects, and gain in terms of documentation readability.

Reply:
Are you advocating for dumbing down the code? Or just the documentation?
<br>
> Furthermore, these are not used anywhere else in FlowCal, and it is very inconsistent to have one single function that suddenly uses this nomenclature.

I agree here. Although I'd prefer improving the robustness everywhere else. There are several places where `isinstance(..., list)` is used, which unnecessarily prevents other sequence data structures (namely `tuple`). The documentation usually specifies `list`, at least, but it still feels unnecessarily restrictive. This would obviously have to be a separate issue, though.

We also appear to use `numpy array` and `array_like` interchangeably in some spots? Which is kinda confusing.

Reply:
> Are you advocating for dumbing down the code? Or just the documentation?

Just for the documentation and error messages. The code looks fine and is probably more robust as you said. We can claim we provide support for the basic data structures and silently provide support for the more general abstract classes.

Reply:
Ahh, OK cool. Yeah, I can get behind that.